File: 2181-merge-nodes-in-between-zeros/2181-merge-nodes-in-between-zeros.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        ListNode* temp = head;
        while (temp) {
            if (temp->next && temp->next->val != 0) {
                temp->val += temp->next->val;
                temp->next = temp->next->next;
            }
            if (temp->next && temp->next->val == 0) {
                temp->next = temp->next->next;
                temp = temp->next;
            }
        }
        return head;
    }
};

File: 2181-merge-nodes-in-between-zeros/2181-merge-nodes-in-between-zeros.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        temp = head
        
        while temp:
            if temp.next and temp.next.val != 0:
                temp.val += temp.next.val
                temp.next = temp.next.next
            if temp.next and temp.next.val == 0:
                temp.next = temp.next.next
                temp = temp.next
        
        return head